lib/Sema/CSDiag.cpp

```diff
@@ -6508,7 +6510,100 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+                         diag::closure_tuple_parameter_destructuring_implicit,
+                         argTupleTy);
+                return true;
+              }
+            }
+
+            auto diag = diagnose(params->getStartLoc(),
+                                 diag::closure_tuple_parameter_destructuring,
+                                 argTupleTy);
+            Type actualArgType;
+            if (auto *actualFnType = CE->getType()->getAs<AnyFunctionType>())
+              actualArgType = actualFnType->getInput();
+
+            auto &sourceMgr = CS->getASTContext().SourceMgr;
+
+            SmallString<64> fixIt;
+            llvm::raw_svector_ostream OS(fixIt);
+
+            // If this is multi-line closure we'd have to insert new lines
+            // in the suggested 'let' to keep the structure of the code intact,
+            // otherwise just use ';' to keep everything on the same line.
+            bool isMultiLineClosure = false;
+            StringRef indent = "";
+            if (auto *body = CE->getBody()) {
+              auto inLine = sourceMgr.getLineNumber(CE->getInLoc());
+              auto bodyLine = sourceMgr.getLineNumber(body->getStartLoc());
+              isMultiLineClosure = bodyLine > inLine;
+              indent =
+                  Lexer::getIndentationForLine(sourceMgr, body->getStartLoc());
+              ;
+            }
+
+            SmallString<16> parameter;
+            llvm::raw_svector_ostream parameterOS(parameter);
+
+            parameterOS << "(";
+            interleave(params->getArray(),
+                       [&](const ParamDecl *param) {
+                         parameterOS << param->getNameStr();
+                       },
+                       [&] { parameterOS << ", "; });
+            parameterOS << ")";
+
+            // If at least one of the parameters has explicitly
+            // declared type let's turn fix-it into N 'let' expressions
+            // to be able to specify it.
+            bool explicitTypes = false;
+            for (auto *param : params->getArray()) {
+              if (!param->getType()->is<TypeVariableType>()) {
+                explicitTypes = true;
+                break;
+              }
+            }
+
+            if (isMultiLineClosure || explicitTypes)
+              OS << '\n' << indent;
+
+            // Let's form 'let <name> : [<type>]? = arg' expression.
+            OS << "let " << parameterOS.str() << " = arg"
+               << (isMultiLineClosure ? "\n" + indent : "; ");
+
+            SmallString<64> argName;
+            llvm::raw_svector_ostream nameOS(argName);
+            if (explicitTypes) {
+              nameOS << "(arg: " << argTupleTy->getString() << ")";
+            } else {
+              nameOS << "(arg)";
+            }
+
+            diag.fixItReplace(params->getSourceRange(), nameOS.str())
+                .fixItInsertAfter(
+                    Lexer::getLocForEndOfToken(sourceMgr, CE->getInLoc()),
```

Comment:
Is it possible to get the start loc of the first thing in the closure, rather than the location of the `in`? Like in the parser that would produce better results.